text/0250-Composite_Samplers.md

```diff
@@ -0,0 +1,337 @@
+
+- ComposableSampler (delegate)
+- maximum sampling (throughput) target rate
+
+The sampler SHOULD measure and keep the average rate of incoming spans, and therefore also of the desired ratio between the incoming span rate to the target span rate.
+Upon invocation of its `GetSamplingAdvice` function, the composite sampler MUST get the `SamplingAdvice` from the delegate sampler, passing the same arguments as received.
+
+The returned `SamplingAdvice` is constructed as follows.
+
+- If using the obtained threshold value as the final threshold would entail sampling more spans than the declared target rate, the sampler SHOULD increase the threshold to a value that would meet the target rate. Several algorithms can be used for threshold adjustment, no particular behavior is prescried by the specification though.
+- The `GetAttributes` function returns the union of the set of Attributes returned by calling the delegate's `GetAttributes` and own Attributes.
+- The `UpdateTraceState` function returns the `Tracestate` as returned by calling `UpdateTraceState` from the delegate's `SamplingAdvice`.
+
+When using `ConsistentRateLimiting` in our requirements example as a replacement for `EachOf` and `RateLimiting`, we are left with no use case for any direct equivalent to `EachOf` in Approach Two.
+However, if such a need arises, `ConsistentEachOf` can be easily defined.
+
+## Summary - Approach Two
+
+### Example - sampling configuration with Approach Two
+
+With the samplers introduced by Approach Two, our example requirements can be coded in a very similar way as with Approach One. However, the work of the samplers configured this way forms a tree of `GetSamplingAdvice` invocations rather than `ShouldSample` invocations as in Approach One.
+
+```
+S = ConsistentRateLimiting(
+     ConsistentAnyOf(
+       ConsistentParentBased(
+         ConsistentRuleBased(ROOT, {
+             (http.target == /healthcheck) => ConsistentAlwaysOff,
+             (http.target == /checkout) => ConsistentAlwaysOn,
+             true => ConsistentFixedThreshold(0.25)
+         }),
+       ConsistentRuleBased(CLIENT, {
+         (http.url == /foo) => ConsistentAlwaysOn
+       }
+     ),
+     1000 * 60
+   )
+```
+
+### Limitations of composite samplers in Approach Two
+
+Making sampling decisions with samplers from Approach Two is more efficient than in Approach One, especially if, platform permitting, `null` values can be used for `GetAttributes` and `UpdateTraceState` functions to represent the prevailing trivial cases of _no-new-attributes_ and _no-special-trace-state-keys_. The only limitation of this approach that it operates exclusively within the domain of Composable Samplers (a subset of Consistent Probability Samplers).
+
+Developers of Composable Samplers should consider that the sampling Decision they advise might be different from the final sampling Decision.
+
+## Prior art
+
+A number of composite samplers are already available as independent contributions
```

Comment:
FYI, I have a few more here (in case you want to link to them) at:
- https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/trace/links-based-sampler
- https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/trace/stratified-sampling-example
- https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/trace/tail-based-sampling-span-level

Reply:
Thanks! Good stuff!